src/backend/distributed/shardsplit/shardsplit_decoder.c

```diff
@@ -90,6 +90,40 @@ replication_origin_filter_cb(LogicalDecodingContext *ctx, RepOriginId origin_id)
 }
 
 
+/*
+ * update_replication_progress is copied from Postgres 15. We use it to send keepalive
+ * messages when we are filtering out the wal changes resulting from the initial copy.
+ * If we do not send out messages long enough, wal reciever will time out.
+ */
+static void
+update_replication_progress(LogicalDecodingContext *ctx, bool skipped_xact)
+{
+	static int changes_count = 0;
+
```

Comment:
If there are parallel shard splits, will this affect the static variable here?
If yes, could we have a hashtable of counts corresponding to the LogicalDecodingContext here?

Reply:
We do not currently support parallel shard splits in a single server. The shared memory structure used by the split decoder is global to all shard split decoders.